src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -632,77 +508,138 @@ public void close() {
-            writer.flush();
+        public void encode(ClusterData clusterData) {
+            this.binaryCodec.writeInt(clusterData.getLane());
+            this.binaryCodec.writeInt(clusterData.getTile());
+            this.binaryCodec.writeInt(clusterData.getX());
+            this.binaryCodec.writeInt(clusterData.getY());
+            this.binaryCodec.writeInt(clusterData.getNumReads());
+
+            for (int i = 0; i < clusterData.getNumReads(); i++) {
+                ReadData read = clusterData.getRead(i);
+                byte[] bases = read.getBases();
+                byte[] quals = read.getQualities();
+
+                binaryCodec.writeInt(bases.length);
+                binaryCodec.writeString(read.getReadType().name(), false, false);
+
+                for (int j = 0; j < bases.length; j++) {
+                    this.binaryCodec.writeByte(bases[j]);
+                    this.binaryCodec.writeByte(quals[j]);
+                }
+            }
+
+            this.binaryCodec.writeBoolean(clusterData.isPf());
+            if (clusterData.getMatchedBarcode() != null) {
+                this.binaryCodec.writeString(clusterData.getMatchedBarcode(), true, true);
+            } else {
+                this.binaryCodec.writeString("", true, false);
+            }
         }
 
-        private void encodeArray(final FastqRecord[] recs) {
-            for (final FastqRecord rec : recs) {
-                writer.write(rec);
+        @Override
+        public ClusterData decode() {
+
+            int lane;
+            try {
+                lane = this.binaryCodec.readInt();
+            } catch (final RuntimeEOFException e) {
+                return null;
+            }
+            int tile = this.binaryCodec.readInt();
+            int x = this.binaryCodec.readInt();
+            int y = this.binaryCodec.readInt();
+
+            int numReads = this.binaryCodec.readInt();
+
+            ReadData[] readData = new ReadData[numReads];
+            for (int i = 0; i < numReads; i++) {
+                int numBases = this.binaryCodec.readInt();
+                String readTypeString = this.binaryCodec.readString(1);
+                ReadType readType = ReadType.valueOf(readTypeString);
+
+                byte[] bases = new byte[numBases];
+                byte[] quals = new byte[numBases];
+
+                for (int j = 0; j < numBases; j++) {
+                    bases[j] = this.binaryCodec.readByte();
+                    quals[j] = this.binaryCodec.readByte();
+                }
+
+                ReadData read = new ReadData();
+                read.setReadType(readType);
+                read.setBases(bases);
+                read.setQualities(quals);
+
+                readData[i] = read;
+            }
+
+            boolean pf = this.binaryCodec.readBoolean();
+            String matchedBarcode = this.binaryCodec.readLengthAndString(false);
+
+            ClusterData clusterData = new ClusterData(readData);
```

Comment:
Why not just declare this at the top of the method and set things directly into it?  That first try/catch could also be simplified then to not have to declare lane outside the try, with:
```java
try { clusterData.setLane(this.binaryCodec.readInt()); }
catch (final RuntimeEOFException e) { return null; }
```

Reply:
I can move it up a bit, but since `ReadData` is private final I need to at least parse `numReads` so I can instantiate the ReadData to set in the constructor.

Alternatively I could open up access to `ReadData` and not have it be final in `ClusterData`

Reply:
```
            int numReads = this.binaryCodec.readInt();

            ReadData[] readData = new ReadData[numReads];
            ClusterData clusterData = new ClusterData(readData);
            clusterData.setLane(this.binaryCodec.readInt());
            clusterData.setTile(this.binaryCodec.readInt());
            clusterData.setX(this.binaryCodec.readInt());
            clusterData.setY(this.binaryCodec.readInt());
            clusterData.setPf(this.binaryCodec.readBoolean());
            String matchedBarcode = this.binaryCodec.readLengthAndString(false);
            if (matchedBarcode.length() == 0) {
                clusterData.setMatchedBarcode(null);
            } else {
                clusterData.setMatchedBarcode(matchedBarcode);
            }
```